flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java

```diff
@@ -0,0 +1,360 @@
+
+	/** Size of network buffer and write buffer. */
+	private final int networkBufferSize;
+
+	/** Current {@link SortBuffer} to append records to. */
+	private SortBuffer currentSortBuffer;
+
+	/** File writer for this result partition. */
+	private PartitionedFileWriter fileWriter;
+
+	public SortMergeResultPartition(
+			String owningTaskName,
+			int partitionIndex,
+			ResultPartitionID partitionId,
+			ResultPartitionType partitionType,
+			int numSubpartitions,
+			int numTargetKeyGroups,
+			int networkBufferSize,
+			ResultPartitionManager partitionManager,
+			FileChannelManager channelManager,
+			@Nullable BufferCompressor bufferCompressor,
+			SupplierWithException<BufferPool, IOException> bufferPoolFactory) {
+
+		super(
+			owningTaskName,
+			partitionIndex,
+			partitionId,
+			partitionType,
+			numSubpartitions,
+			numTargetKeyGroups,
+			partitionManager,
+			bufferCompressor,
+			bufferPoolFactory);
+
+		this.channelManager = checkNotNull(channelManager);
+		this.networkBufferSize = networkBufferSize;
+		this.numDataBuffers = new int[numSubpartitions];
+		this.writeBuffer = MemorySegmentFactory.allocateUnpooledOffHeapMemory(networkBufferSize);
+	}
+
+	@Override
+	protected void releaseInternal() {
+		synchronized (lock) {
+			isFinished = true; // to fail writing faster
+
+			// delete the produced file only when no reader is reading now
+			if (readers.isEmpty()) {
+				if (resultFile != null) {
+					resultFile.deleteQuietly();
+					resultFile = null;
+				}
+			}
+		}
+	}
+
+	@Override
+	public void emitRecord(ByteBuffer record, int targetSubpartition) throws IOException {
+		emit(record, targetSubpartition, DataType.DATA_BUFFER);
+	}
+
+	@Override
+	public void broadcastRecord(ByteBuffer record) throws IOException {
+		broadcast(record, DataType.DATA_BUFFER);
+	}
+
+	@Override
+	public void broadcastEvent(AbstractEvent event, boolean isPriorityEvent) throws IOException {
+		Buffer buffer = EventSerializer.toBuffer(event, isPriorityEvent);
+		try {
+			ByteBuffer serializedEvent = buffer.getNioBufferReadable();
+			broadcast(serializedEvent, buffer.getDataType());
+		} finally {
+			buffer.recycleBuffer();
+		}
+	}
+
+	private void broadcast(ByteBuffer record, DataType dataType) throws IOException {
+		for (int channelIndex = 0; channelIndex < numSubpartitions; ++channelIndex) {
+			record.rewind();
+			emit(record, channelIndex, dataType);
+		}
+	}
+
+	private void emit(ByteBuffer record, int targetSubpartition, DataType dataType) throws IOException {
+		checkInProduceState();
+
+		SortBuffer sortBuffer = getSortBuffer();
+		if (sortBuffer.append(record, targetSubpartition, dataType)) {
+			return;
+		}
+
+		if (!sortBuffer.hasRemaining()) {
+			// the record can not be appended to the free sort buffer because it is too large
+			releaseCurrentSortBuffer();
+			writeLargeRecord(record, targetSubpartition, dataType);
+			return;
+		}
+
+		flushCurrentSortBuffer();
+		emit(record, targetSubpartition, dataType);
```

Comment:
I think the below logic is not very easily to understand and also has some overheads in practice. 

```
if (!sortBuffer.hasRemaining()) {
	// the record can not be appended to the free sort buffer because it is too large
	releaseCurrentSortBuffer();
        writeLargeRecord(record, targetSubpartition, dataType);
	return;
}

flushCurrentSortBuffer();
emit(record, targetSubpartition, dataType);
```

Here we have to flush current buffer and call `emit` again to try out whether the unfilled sort buffer can hold the large record or not. During `sortBuffer.append` it would request all the segments from `LocalBufferPool` until exhausted before return `false` to decide flush large record.

I think of two options for improvement here.

- Make `SortMergeResultPartition` more light-weight component to only dispatch record/events, then the `SortBuffer` can decide itself when to append records in cache, when to flush them and when to flush large records directly internally. No need to expose many interfaces to interact with `SortMergeResultPartition`.

- `SortBuffer` provide another method to expose the cache threshold, it can be easily got from the internal `LocalBufferPool`. Then `SortMergeResultPartition` can decide whether to flush the large record directly to avoid unnecessary requesting all the segments.

Reply:
For the first option, the change is a little big and I would like to treat SortBuffer a purely in-memory data structure.
For the second option, the current size of buffer pool can be changed if new buffer pool created or old buffer pool is destroyed.

Reply:
Thanks for the explanation. I might overlook the case of dynamic buffer pool for the second option.
I am not quite sure whether it is worth improving the overhead or simplifying the logic ATM. Maybe we can later decide it if necessary future.